### DISCUSSION TITLE
Why sometimes browser can return lowdefy.contexts and sometimes can not?

### DISCUSSION THREAD

**Discussion by 13316090972** [lowdefy/lowdefy#1024] [General]
I open two browser tab,  one visit the app local,the other visit the app remote.
The first tab can return the value of lowdefy.contexts on the browser console,the second just return undefined.
What's the reason?
![image](https://user-images.githubusercontent.com/56707831/145510265-a9d3d7f4-0c1a-477d-aded-aa6d97e32340.png)
![image](https://user-images.githubusercontent.com/56707831/145510301-bf8a631e-9404-4868-a861-e5ab83fa7472.png)

**Comment by SamTolmay** [comment]
Hi @13316090972 

We have done this since we want the lowdefy object to only be visible when developing the app, but not when running in production. To do this, we test if the app is running at `http://localhost`, and then add the `lowdefy` to the global scope.

This will change slightly in version in 4, as we will have a development server, which will add `lowdefy` to the global scope, and a production server which won't.

**Reply by 13316090972** [reply]
Do you please to give some more detail for "and then add the lowdefy to the global scope"?

**Reply by SamTolmay** [reply]
Javascript has a `global` object, which in the case of the browser is called `window`. You can read more about it [here](https://developer.mozilla.org/en-US/docs/Glossary/Global_object). If we add lowdefy to the window object (`window.lowdefy = lowdefy`), then you can easily view that object by typing `lowdefy` in the browser menu. 

I one thing I missed in my previous explanation is that there is always a `lowdefy` object on `window`, the one in production just has less data inside it.

The relevent code is at line 51 here: https://github.com/lowdefy/lowdefy/blob/main/packages/renderer/src/Renderer.js
